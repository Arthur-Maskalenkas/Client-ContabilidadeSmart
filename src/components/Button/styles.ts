import styled, { css, DefaultTheme } from 'styled-components'
import { darken } from 'polished'

import { ButtonProps } from '.'

// hasIcon Ã© criado e passado diretamente dentro do wrapper, por isso esta sendo pego aqui
// Esta pegando as {...props} implicitamente
export type WrapperProps = {
  hasIcon: boolean
} & Pick<ButtonProps, 'size' | 'fullWidth'>

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  fullWidth: () => css`
    width: 100%;
  `,
  withIcon: (theme: DefaultTheme) => css`
    svg {
      width: 1.5rem;
      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `,

  disabled: () => css`
    &:disabled {
      cursor: not-allowed;
      filter: saturate(30%);
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, fullWidth, hasIcon, disabled }) => css`
    display: inline-flex;
    color: ${darken(0.4, theme.colors.primary)};
    align-items: center;
    justify-content: center;
    background: ${theme.colors.primary};
    font-family: ${theme.font.family};
    border: 0;
    cursor: pointer;
    border-radius: ${theme.border.radius};
    padding: ${theme.spacings.xxsmall};
    text-decoration: none;
    transition: 0.3s all;
    &:hover {
      transition: 0.3s all;
      background: ${theme.colors.secondary};
    }
    ${!!size && wrapperModifiers[size](theme)};
    ${!!fullWidth && wrapperModifiers.fullWidth()};
    ${!!hasIcon && wrapperModifiers.withIcon(theme)};
    ${!!disabled && wrapperModifiers.disabled()};
  `}
`
