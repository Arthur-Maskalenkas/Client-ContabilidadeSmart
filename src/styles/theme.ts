export default {
  grid: {
    container: '130rem',
    gutter: '1.5rem'
  },
  border: {
    radius: '0.4rem'
  },
  font: {
    family: {
      poppins: {
        font: "Poppins, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
        weight: {
          light: 300,
          normal: 400,
          bold: 600
        }
      },
      raleway: {
        font: "Raleway, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
        weight: {
          light: 300,
          normal: 400,
          bold: 800
        }
      }
    },
    sizes: {
      xsmall: '1.2rem',
      small: '2rem',
      medium: '2.8rem',
      large: '3.6rem',
      xlarge: '4.2rem',
      xxlarge: '5rem',
      huge: '7rem'
    }
  },
  colors: {
    primary: '#FFFFFF',
    secondary: '#F66A3E',
    tertiary: '#202020',
    lightBg: '#F2F2F2',
    mainBg: '#F8F8F8',
    artigosBg: '#FBFBFB',
    white: '#FFFFFF',
    black: '#050706',
    lightGray: '#EFEFEF',
    gray: '#8F8F8F',
    darkGray: '#3F4350',
    red: '#FF6347',
    yellow: 'F2BB05',
    green: '#32A09A',
    lightGreen: '#80CFCF',
    darkGreen: '#278282',
    brown: '#1d1a1a',
    bordasCinzas: '#535662'
  },
  spacings: {
    xxxsmall: '0.4rem',
    xxsmall: '0.8rem',
    xsmall: '1.6rem',
    small: '2.4rem',
    medium: '3.2rem',
    large: '4.0rem',
    xlarge: '4.8rem',
    xxlarge: '5.6rem'
  },
  layers: {
    base: 10,
    menu: 20,
    overlay: 30,
    modal: 40,
    alwaysOnTop: 50
  },
  transition: {
    default: '0.3s ease-in-out',
    fast: '0.1s ease-in-out'
  },
  shadow: {
    card: '0px 0px 24px -16px #000000'
  }
} as const
