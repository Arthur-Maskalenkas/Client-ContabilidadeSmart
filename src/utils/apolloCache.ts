import { InMemoryCache } from '@apollo/client'
import { concatPagination } from '@apollo/client/utilities'

// A query QueryGames retorna -> games <-, e é isso que o cache vai acumular games

// ** fluxo:
// > Eu pego uma lista de free (retorna games assim como varias outras query)
// > o cache fica [game1,game2,game3]
// > Assim que outra parte usa alguma query que retorne game, ela vai receber independente de tudo [game1,game2,game3], e não [game4,game5,game6]
// **

// Para quebrar o fluxo, eu separei o cacheamento por 'where' e 'sort'. Assim fica algo como: (caso puxe 3 games de 3 queries diferentes)

//**
// > games: {"where: {"release_date_lte":"2021-01-29"}, "sort":"release_date:desc"}: [] 8 items"}
// > games: {"where: {"release_date_gt":"2021-01-29"}, "sort":"release_date:asc"}: [] 8 items"}
// > games: {"where: {"price": 0}}: [] 8 items
// > Antes ficaria tudo em games com 24 items
//**
export default new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        posts: concatPagination(['where', 'sort'])
      }
    }
  }
})
