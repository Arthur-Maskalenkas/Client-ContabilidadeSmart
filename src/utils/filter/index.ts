import { ParsedUrlQueryInput } from 'querystring'

type ParseArgs = {
  queryString: ParsedUrlQueryInput
}

export const parseQueryStringToWhere = ({ queryString }: ParseArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {}

  // queryString = {platforms: ['windows','linux']}
  // filterItems = {name: 'platforms', type: 'checkbox'}
  Object.keys(queryString).forEach((key) => {
    // Pegando o valor e colocando dentro do objeto
    obj[key] = { slug_contains: queryString[key] }
  })

  return obj
}
