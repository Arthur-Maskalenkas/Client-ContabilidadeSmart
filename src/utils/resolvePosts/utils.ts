export const operations = {
  nextPage: nextPageOperation(),
  backPage: backPageOperation()
}

type variablesType = {
  limit?: number
  start?: number
}

export type FetchMoreProps = {
  variables: variablesType
}
