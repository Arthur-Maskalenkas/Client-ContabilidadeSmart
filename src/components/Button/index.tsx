import { AnchorHTMLAttributes, ButtonHTMLAttributes, forwardRef } from 'react'
import * as S from './styles'

//https://www.saltycrane.com/cheat-sheets/typescript/react/latest/

// Componente pode receber href, onclick...
type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  fullWidth?: boolean
  icon?: JSX.Element
  as?: React.ElementType
} & ButtonTypes

// ...props vai ter as, que por sua vez pode ter a e o restante dos atribtos dele
/* href tem que chegar em um a. Ele vai para o <button as="a"> > vai passar o href para o wrapper que por sua vez vai receber um as.*/
// Esta retornando um forward ref

// FUnciona assim:
//<Link href="/sign-in" passHref>
// <Button as="a">Sign in</Button>
//</Link>

//Depois disso o wrapper vai receber essa Href

// O passHref vai chegar aqui e vai ser tratado, jogando para o wrapper que Ã© um a.
const Button: React.ForwardRefRenderFunction<S.WrapperProps, ButtonProps> = (
  { children, icon, size = 'medium', fullWidth = false, ...props },
  ref
) => (
  <S.Wrapper
    size={size}
    fullWidth={fullWidth}
    hasIcon={!!icon}
    ref={ref}
    {...props}
  >
    {icon}

    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default forwardRef(Button)
