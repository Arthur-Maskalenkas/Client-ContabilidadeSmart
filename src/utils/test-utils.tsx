import { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'

import { ThemeProvider } from 'styled-components'
import theme from 'styles/theme'

type CustomRenderProps = Omit<RenderOptions, 'queries'>

const customRender = (ui: ReactElement, { ...renderOptions }: CustomRenderProps = {}) =>
  render(<ThemeProvider theme={theme}>{ui}</ThemeProvider>, renderOptions)

// Tudo que eu chamar, vai vim dessa pasta aqui
export * from '@testing-library/react'
export { customRender as render }
