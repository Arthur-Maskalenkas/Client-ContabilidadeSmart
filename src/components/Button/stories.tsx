import { Story, Meta } from '@storybook/react/types-6-0'
import Button from '.'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    }
  }
} as Meta

export const Default: Story = (args) => <Button {...args} />

Default.args = {
  children: 'Buy now'
}

export const withIcon: Story = (args) => <Button {...args} />

export const asLink: Story = (args) => <Button {...args} />

asLink.args = {
  size: 'large',
  children: 'Buy now',
  as: 'a',
  href: '/link'
}

withIcon.args = {
  size: 'medium',
  children: 'Buy now',
  icon: <AddShoppingCart />
}
