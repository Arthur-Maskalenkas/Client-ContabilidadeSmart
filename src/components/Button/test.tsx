import { ArrowRight } from '@styled-icons/material-outlined'
import theme from 'styles/theme'
import { render, screen } from 'utils/test-utils'

import Button from '.'

describe('<Button />', () => {
  it('isso deve renderizar o botão', () => {
    render(<Button>clique aqui</Button>)

    const button = screen.getByText(/clique aqui/i)

    expect(button).toBeInTheDocument()
    expect(button).toHaveStyle({
      color: theme.colors.primary
    })
    expect(button).toHaveStyle({
      backgroundColor: theme.colors.secondary
    })
  })

  it('isso deve renderizar o botão com o icone', () => {
    render(
      <Button icon={<ArrowRight title="arrow Right" />}>clique aqui</Button>
    )

    const button = screen.getByText(/clique aqui/i)

    expect(button).toBeInTheDocument()
    expect(screen.getByTitle(/arrow right/i)).toBeInTheDocument()
  })

  it('vai ter link', () => {
    render(
      <Button as="a" href="/link">
        Buy now
      </Button>
    )

    expect(screen.getByRole('link', { name: /buy now/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })

  describe('<Button/> tamanhos', () => {
    it('vai renderizar com o tamanho small', () => {
      render(<Button size="small">Buy now</Button>)

      const button = screen.getByRole('button', { name: /Buy now/i })

      expect(button).toHaveStyle({
        height: '3rem',
        'font-size': '1.2rem'
      })
    })

    it('vai renderizar com o tamanho grande', () => {
      render(<Button size="large">Buy now</Button>)

      const button = screen.getByRole('button', { name: /Buy now/i })

      expect(button).toHaveStyle({
        height: '5rem',
        'font-size': '2.8rem',
        padding: '0.8rem 4.8rem'
      })
    })
  })
})
