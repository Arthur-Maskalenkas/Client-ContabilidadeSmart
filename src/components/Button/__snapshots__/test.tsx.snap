// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size by default 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  color: #FAFAFA;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #F51700;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: 0.3s all;
  transition: 0.3s all;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c0:hover {
  -webkit-transition: 0.3s all;
  transition: 0.3s all;
  background: #c21200;
}

<button
  class="c0"
>
  <span>
    Buy now
  </span>
</button>
`;
